Add empty constructor for Linear DS

domire8 opened the pull request:
First of all, for reasons that I don't know, `Parameter<CartesianPose>("attractor", CartesianPose()))` sets the CartesianPose as filled, even though its empty. This is why I set it back to empty manually. 
The base frame of the DS is set to the identity in the base DS class, that's why this doesn't need special care. Instead, the user is required to set at least the attractor such that it will be possible to use the `evaluate` function (which won't be meaningful until the gain is set from different from 0..).

Let's hear your opinion, we could also leave the attractor as filled and only set the gain to zero to avoid the user from using the DS after an empty construction without modifying the attractor/base_frame fields

eeberhard commented on source/dynamical_systems/include/dynamical_systems/exceptions/AttractorEmptyException.hpp:
```diff
@@ -0,0 +1,11 @@
+#include <exception>
+
+namespace dynamical_systems::exceptions {
+class AttractorEmptyException : public std::runtime_error {
```
It's a small thing, but I think `EmptyAttractorException` would be a bit nicer (like IncompatibleSize, it's putting the adjective first). Plus then it forms my initials EAE hehe

eeberhard reviewed (commented):
I think this is good. The emptiness of derived classes like Linear will be more complicated than the base class, so it makes sense to add an extra layer of checking if the attractor is empty.

But, I think a good place to start is to also make the base DynamicalSystem base_frame_ empty in the default constructor (e.g. by default / empty constructing the state too). You can then add one more exception EmptyBaseFrameException and check that at the top of `DynamicalSystem::evaluate`.

Finally, you can link some behaviours - for example, in Linear::set_attractor, you might want to initialise the base frame properly if it is also empty. Making these changes to the base class and implementing them in Linear would set a nice pattern I think.

domire8 commented:
What is "intialize the base frame properly" for you @eeberhard ? Set it according to the attractor reference frame, e.g. identity but with attractor-compatible name/reference frame?

eeberhard commented:
> What is "intialize the base frame properly" for you @eeberhard ? Set it according to the attractor reference frame, e.g. identity but with attractor-compatible name/reference frame?

Exactly, similarly to how it is constructed in the Linear DS: identity with the same name and reference frame as the attractor reference frame

domire8 commented:
Okay so up to now, and this might be a good point for a first review, I changed the following things:
- In the base class `DynamicalSystem`, the empty constructor now initializes with an empty state in all cases. This requires a check in the `evaluate` function for the emptiness of the base_frame.
- Added an empty constructor for the `Linear<CartesianState>`. This requires checking for an empty attractor in the `compute_dynamics` function. 
- Additionally, if you want to `set_base_frame` or `set_attractor`, or simply construct the `Linear` with an attractor, the provided state cannot be empty, otherwise an exception is thrown.
- Last but not least, from @eeberhard suggestions, if you set the attractor and the base_frame is empty, the base_frame is set to identity with name and reference_frame corresponding to the attractors reference_frame. This does not happen for the other way round, i.e. when the base_frame is set the attractor stays empty

domire8 commented:
> Okay so up to now, and this might be a good point for a first review, I changed the following things:
> 
>     * In the base class `DynamicalSystem`, the empty constructor now initializes with an empty state in all cases. This requires a check in the `evaluate` function for the emptiness of the base_frame.
> 
>     * Added an empty constructor for the `Linear<CartesianState>`. This requires checking for an empty attractor in the `compute_dynamics` function.
> 
>     * Additionally, if you want to `set_base_frame` or `set_attractor`, or simply construct the `Linear` with an attractor, the provided state cannot be empty, otherwise an exception is thrown.
> 
>     * Last but not least, from @eeberhard suggestions, if you set the attractor and the base_frame is empty, the base_frame is set to identity with name and reference_frame corresponding to the attractors reference_frame. This does not happen for the other way round, i.e. when the base_frame is set the attractor stays empty
> 
> 
> I also tried to cover this by unittests as good as possible.

Also added
- Empty constructor for the `Linear<JointState>` DS and similar changes to above. It might be worth taking some times to check if my interpretation of the base frame makes sense.
- Improve / Add unittests for `Linear<JointState>`

I also had to move around some functions from source file to header file, I don't really get whats going on with this templated class but its how it compiles.

buschbapti commented on source/dynamical_systems/include/dynamical_systems/Linear.hpp:
```diff
@@ -16,6 +16,8 @@
 #include "state_representation/space/cartesian/CartesianState.hpp"
 #include "state_representation/space/cartesian/CartesianTwist.hpp"
 
+using namespace state_representation;
```
I am not certain we want that in header files. In source file this is not a problem because it can't break much but in header files this basically mean that any file that includes the header will also use this namespace. I prefer that we use a verbose solution than to do that.

buschbapti commented on source/dynamical_systems/include/dynamical_systems/Linear.hpp:
```diff
@@ -37,29 +39,34 @@ class Linear : public DynamicalSystem<S> {
    * @param iso_gain the iso gain of the system
    */
-  explicit Linear(const S& attractor, double iso_gain = 1.0);
+  Linear(const S& attractor, double iso_gain = 1.0);
```
Why did you remove the `explicit` for those? I think for the empty constructor it makes sense but here they should be kept

domire8 replied:
@eeberhard advised me to remove `explicit` for the Jacobian constructors too, which have more than one argument. I assumed this is applicable here too

eeberhard replied:
It's a little different here because this can be treated as a single argument constructor (because of the default value). So, explicit does have a purpose - if you keep it, then the passed attractor type _must_ be the same as the class template type. Without explicit, you could do `Linear<CartesianState>(CartesianPose::Identity(...))`. So the question is more about which behaviour you want. The rule of thumb is to remove `explicit` from constructors that have 0 or >1 argument (excluding default arguments), and to think about the desired behaviour single-argument cases.

buschbapti replied:
I am pretty sure I was using `Linear<CartesianState>(CartesianPose)` even with explicit. But probably because a `CartesianPose` is also a `CartesianState`. To be honest I never really know what are the actual implications for it but I am happy to have some rules that we always apply

buschbapti commented on source/dynamical_systems/include/dynamical_systems/Linear.hpp:
```diff
@@ -107,9 +114,15 @@ class Linear : public DynamicalSystem<S> {
 };
 
+template<>
+void Linear<JointState>::set_gain(double iso_gain) {
```
if you put a specialized version in the header then it should be because you want to `inline` it. Otherwise it should go in the source file.

domire8 replied:
I needed to put it here because it's used in the subsequent `set_attractor` for the `JointState`. But I forgot to `inline` it

buschbapti commented on source/dynamical_systems/src/DynamicalSystem.cpp:
```diff
@@ -31,15 +32,30 @@ bool DynamicalSystem<state_representation::CartesianState>::is_compatible(const
 
+template<>
+bool DynamicalSystem<state_representation::JointState>::is_compatible(const state_representation::JointState& state) {
+  return (state.get_name() == this->get_base_frame().get_name()
```
Does this makes sense for `JointState`? Because it means you need to have a base frame which I am not sure what would be the usage. I think it actually would make sense to check compatibility with the attractor, in that case putting the function `virtual` and `overriding` it in the base classes. You could also use directly the `is_compatible` function of `JointState` actually.

buschbapti replied:
I guess if your derived class constructors correctly initialize it then it should be okay. It would not exactly be a "base frame" in that case but more like an accepted state.

domire8 replied:
Good point and one that was destined to come up. Indeed for the JointState it doesn't really make sense, I just though of it as a place holder for the name and joint_names because the base_frame_ is required to be set in the base class. The derived class constructors initialize them with the same name and joint_names as the attractor, but it could technically be changes by the user with `set_base_frame`. So I'm not sure what the best solution is here.

eeberhard replied:
I think it makes sense - while we don't have the same reference frame logic as in Cartesian space, we still expect the robot and joint names to be the same in order to be compatible. The base frame still has a meaning in this case, and it's up to the derived classes to set it according to the intended usage.

buschbapti reviewed (commented):
This seems good to me. I just made a few comments.

eeberhard commented on source/dynamical_systems/test/tests/test_linear.cpp:
```diff
@@ -1,51 +1,125 @@
+  EXPECT_TRUE(ds.get_attractor().is_empty());
+  ds.set_attractor(attractor);
+  EXPECT_FALSE(ds.get_attractor().is_empty());
+  // when attractor was set without a base frame, expect base frame to be identity with name / reference_frame of attractor
```
I guess it's implied by the subsequent expects, but it might be nice to have an `EXPECT_FALSE(get_base_frame().is_empty());` here for explicitness. (same for JointState test)

eeberhard commented on source/dynamical_systems/src/DynamicalSystem.cpp:
```diff
@@ -31,15 +32,30 @@ bool DynamicalSystem<state_representation::CartesianState>::is_compatible(const
+JointState DynamicalSystem<JointState>::evaluate(const JointState& state) const {
+  if (this->get_base_frame().is_empty()) {
+    throw exceptions::EmptyBaseFrameException("The base frame of the dynamical system is empty.");
+  }
```
I would add the compatibility check here, similar to CartesianState evaluate. The difference is that the Cartesian one has some intermediate logic so can't call `is_compatible` directly, but we can do it here.

```suggestion
  }
  if (!this->is_compatible(is_compatible)) {
    state_representation::exceptions::IncompatibleReferenceFramesException(...);
  }
```

domire8 replied:
Yes that makes sense, more consistency thanks!

eeberhard reviewed (commented):
This is some thorough work and represents a generally useful design improvement. It only affects users who want to empty construct objects instead of deferring construction, which makes certain demos and cases more readable / less verbose, and tailors nicely towards new users. In exchange, the source code becomes a bit more verbose, but that's a fair trade-off for the improved usage flexibility.

Just some minor comments, then I think it's good to go

buschbapti commented:
All makes a lot of sense. This is good. Thanks for that.

domire8 commented on source/dynamical_systems/src/DynamicalSystem.cpp:
```diff
@@ -14,39 +15,57 @@ template<>
 
 template<class S>
-bool DynamicalSystem<S>::is_compatible(const S&) {
+bool DynamicalSystem<S>::is_compatible(const S&) const {
```
Ah i just saw that this function is `virtual`, should I `override` it?

eeberhard replied:
well no, because it's declared virtual in this class, there is no ancestor to override!

domire8 commented:
Thanks guys, merging now
